ydb/core/tablet_flat/flat_part_charge_btree_index.h

```diff
@@ -11,56 +11,110 @@ class TChargeBTreeIndex : public ICharge {
+    using TChild = TBtreeIndexNode::TChild;
+
+    // TODO: store PageId only instead of TChild?
+    struct TNodeState : TBtreeIndexNode, TChild {
+        TRowId BeginRowId;
+        TRowId EndRowId;
+
+        TNodeState(TSharedData data, TChild meta, TRowId beginRowId, TRowId endRowId)
+            : TBtreeIndexNode(data) 
+            , TChild(meta)
+            , BeginRowId(beginRowId)
+            , EndRowId(endRowId)
+        {
+        }
+    };
 
 public:
     TChargeBTreeIndex(IPages *env, const TPart &part, TTagsRef tags, bool includeHistory = false)
         : Part(&part)
+        , Scheme(*Part->Scheme)
         , Env(env) {
-        Y_UNUSED(env);
         Y_UNUSED(part);
         Y_UNUSED(tags);
         Y_UNUSED(includeHistory);
     }
 
 public:
-    TResult Do(const TCells key1, const TCells key2, TRowId row1, TRowId row2, 
+    TResult Do(TCells key1, TCells key2, TRowId row1, TRowId row2, 
             const TKeyCellDefaults &keyDefaults, ui64 itemsLimit, ui64 bytesLimit) const noexcept override {
         bool ready = true;
 
-        Y_UNUSED(key1);
-        Y_UNUSED(key2);
-        Y_UNUSED(keyDefaults);
         Y_UNUSED(itemsLimit);
         Y_UNUSED(bytesLimit);
 
         const auto& meta = Part->IndexPages.BTreeGroups[0];
 
+        if (meta.LevelCount == 0) {
+            ready &= HasDataPage(meta.PageId, { });
+            return { ready, true };
+        }
+
         if (Y_UNLIKELY(row1 >= meta.RowCount)) {
             return { true, true }; // already out of bounds, nothing to precharge
         }
         if (Y_UNLIKELY(row1 > row2)) {
             row2 = row1; // will not go further than row1
         }
+        TRowId sliceRow2 = row2;
+        if (Y_UNLIKELY(key1 && key2 && Compare(key1, key2, keyDefaults) > 0)) {
+            key2 = key1; // will not go further than key1
+        }
 
-        TVector<TBtreeIndexNode> level, nextLevel(Reserve(2));
-        for (ui32 height = 0; height < meta.LevelCount && ready; height++) {
-            if (height == 0) {
-                ready &= TryLoadNode(meta.PageId, nextLevel);
-            } else {
-                for (ui32 i : xrange<ui32>(level.size())) {
-                    TRecIdx from = 0, to = level[i].GetKeysCount();
-                    if (i == 0) {
-                        from = level[i].Seek(row1);
-                    }
-                    if (i + 1 == level.size() && row2 < meta.RowCount) {
-                        to = level[i].Seek(row2);
-                    }
-                    for (TRecIdx j : xrange(from, to + 1)) {
-                        ready &= TryLoadNode(level[i].GetShortChild(j).PageId, nextLevel);
+        TVector<TNodeState> level(Reserve(3)), nextLevel(Reserve(3));
+        TPageId key1PageId = key1 ? meta.PageId : Max<TPageId>();
+        TPageId key2PageId = key2 ? meta.PageId : Max<TPageId>();
+
+        const auto iterateLevel = [&](std::function<bool(TNodeState& current, TRecIdx pos)> tryLoadNext) {
```

Comment:
I would recommend making the argument `const auto& tryLoadNext` here. You are passing lambdas below, and packaging them into std::function may be expensive or even allocate memory depending on state size.